director.go

```diff
@@ -0,0 +1,200 @@
+	}
+
+	return retMap
+}
+
+// Given the Director response, create the ordered list of caches
+// and store it as namespace.SortedDirectorCaches
+func CreateNSFromDirectorResp(dirResp *http.Response, namespace *Namespace) (err error) {
+	X_OSDF_Namespace := HeaderParser(dirResp.Header.Values("X-Osdf-Namespace")[0])
+	namespace.Path = X_OSDF_Namespace["namespace"]
+	namespace.UseTokenOnRead, _ = strconv.ParseBool(X_OSDF_Namespace["use-token-on-read"])
+	namespace.ReadHTTPS, _ = strconv.ParseBool(X_OSDF_Namespace["readhttps"])
+
+	var X_OSDF_Authorization map[string]string
+	if len(dirResp.Header.Values("X-Osdf-Authorization")) > 0 {
+		X_OSDF_Authorization = HeaderParser(dirResp.Header.Values("X-Osdf-Authorization")[0])
+		namespace.Issuer = X_OSDF_Authorization["issuer"]
+	}
+
+	// Create the caches slice
+	namespace.SortedDirectorCaches, err = GetCachesFromDirectorResponse(dirResp)
+	if err != nil {
+		log.Errorln("Unable to construct ordered cache list:", err)
+		return
+	}
+	log.Debugln("Namespace path constructed from Director:", namespace.Path)
+
+	return
+}
+
+func QueryDirector(source string, directorUrl string) (resp *http.Response, err error) {
+	resourceUrl := directorUrl + source
+
+	// Prevent following the Director's redirect
+	client := &http.Client{
+		CheckRedirect: func(req *http.Request, via []*http.Request) error {
+			return http.ErrUseLastResponse
+		},
+	}
+
+	log.Debugln("Querying OSDF Director at", resourceUrl)
+	resp, err = client.Get(resourceUrl)
+	log.Debugln("Director's response:", resp)
+
+	if err != nil {
+		log.Errorln("Failed to get response from OSDF Director:", err)
+		return
+	}
+
+	defer resp.Body.Close()
+	return
+}
+
+func GetCachesFromDirectorResponse(resp *http.Response) (caches []DirectorCache, err error) {
+	// Get the Link header
+	linkHeader := resp.Header.Values("Link")
+
+	for _, linksStr := range strings.Split(linkHeader[0], ",") {
+		links := strings.Split(strings.ReplaceAll(linksStr, " ", ""), ";")
+
+		var endpoint string
+		// var rel string // "rel", as defined in the Metalink/HTTP RFC. Currently not being used by
+		// the OSDF Client, but is provided by the director.
+		var pri int
+		for _, val := range links {
+			if strings.HasPrefix(val, "<") {
+				endpoint = val[1 : len(val)-1]
+			} else if strings.HasPrefix(val, "pri") {
+				pri, _ = strconv.Atoi(val[4:])
+			}
+			// } else if strings.HasPrefix(val, "rel") {
+			// 	rel = val[5 : len(val)-1]
+			// }
+		}
+
+		// Construct the cache objects, populating only the url+port that will be used
+		// based on authentication. Also, cache.Resource is currently being set as
+		// the priority, because the Director at this time doesn't provide a resource
+		// name. Maybe there's a way to bake that into the LINK header for each cache
+		// while still following Metalink/HTTP?
+		var cache DirectorCache
+		port := strings.Split(endpoint, ":")[1]
+		if port == "8000" {
+			cache.AuthedReq = false
+		} else if port == "8443" {
```

Comment:
At some point, we may want TLS on every transfer, even if they are not authenticated.  But this is probably a fine assumption now, it won't be in the future.  Where is cache.AuthedReq used later?